File: JsBridgeDemo/WebViewJavascriptBridge/BkJsBridge.m
//
//  WebViewJavascriptBridge.m
//  ExampleApp-iOS
//
//  Created by Marcus Westin on 6/14/13.
//  Copyright (c) 2013 Marcus Westin. All rights reserved.
//

#import "BkJsBridge.h"

@implementation BkJsBridge
{
    UIWebView *_webView;
    NSMutableDictionary *_responseCallbackDic;
    NSMutableDictionary *_messageHandlerDic;
    long _uniqueId;
    NSUInteger _numRequestsLoading;
}

#pragma mark - public
//初始化
- (id)initBridgeForWebView:(UIWebView *)webView
{
    self = [super init];
    if (self)
    {
        _webView = webView;
        _webView.delegate = self;
        _messageHandlerDic = [[NSMutableDictionary alloc] init];
        _responseCallbackDic = [[NSMutableDictionary alloc] init];
        _uniqueId = 0;
    }
    
    return self;
}

//注册oc方法，供js调用
- (void)registerHandler:(NSString *)handlerName handler:(WVJBHandler)handler
{
    [_messageHandlerDic setObject:handler forKey:handlerName];
}

//oc调用 js方法
- (void)callHandler:(NSString *)handlerName data:(id)data responseCallback:(WVJBResponseCallback)responseCallback
{
    NSMutableDictionary *message = [[NSMutableDictionary alloc] init];
    if (data)
    {
        [message setObject:data forKey:@"data"];
    }
    
    if (responseCallback)
    {
        NSString *callbackId = [NSString stringWithFormat:@"objc_cb_%ld", ++_uniqueId];
        [_responseCallbackDic setObject:responseCallback forKey:callbackId];
        [message setObject:callbackId forKey:@"callbackId"];
    }
    
    if (handlerName)
    {
        [message setObject:handlerName forKey:@"handlerName"];
    }
    
    [self queueMessage:message];
}

#pragma mark - private
//运行消息队列
- (void)queueMessage:(NSDictionary *)message
{
    //序列化JSON
    NSString *messageJSON = [self serializeMessage:message];
    
    messageJSON = [messageJSON stringByReplacingOccurrencesOfString:@"\\" withString:@"\\\\"];
    messageJSON = [messageJSON stringByReplacingOccurrencesOfString:@"\"" withString:@"\\\""];
    messageJSON = [messageJSON stringByReplacingOccurrencesOfString:@"\'" withString:@"\\\'"];
    messageJSON = [messageJSON stringByReplacingOccurrencesOfString:@"\n" withString:@"\\n"];
    messageJSON = [messageJSON stringByReplacingOccurrencesOfString:@"\r" withString:@"\\r"];
    messageJSON = [messageJSON stringByReplacingOccurrencesOfString:@"\f" withString:@"\\f"];
    
    NSString *javascriptCommand = [NSString stringWithFormat:@"BkJsBridge.handleMessageFromObjC('%@');", messageJSON];
    if ([[NSThread currentThread] isMainThread])
    {
        [_webView stringByEvaluatingJavaScriptFromString:javascriptCommand];
    }
}

//刷新消息队列
- (void)refreshMessageQueue
{
    NSString *messageQueueString = [_webView stringByEvaluatingJavaScriptFromString:@"BkJsBridge.fetchQueue();"];
    
    //反序列化JSON
    id messages = [self deserializeMessageJSON:messageQueueString];
    if (![messages isKindOfClass:[NSArray class]])
    {
        return;
    }
    
    for (NSDictionary *message in messages)
    {
        if (![message isKindOfClass:[NSDictionary class]])
        {
            continue;
        }
        
        NSString *responseId = [message objectForKey:@"responseId"];
        if (responseId)
        {
            WVJBResponseCallback responseCallback = [_responseCallbackDic objectForKey:responseId];
            responseCallback([message objectForKey:@"responseData"]);
            [_responseCallbackDic removeObjectForKey:responseId];
        }
        else
        {
            WVJBResponseCallback responseCallback = NULL;
            NSString *callbackId = [message objectForKey:@"callbackId"];
            if (callbackId)
            {
                responseCallback = ^(id responseData)
                {
                    NSDictionary* msg = @{ @"responseId":callbackId, @"responseData":responseData };
                    [self queueMessage:msg];
                };
            }
            else
            {
                responseCallback = ^(id ignoreResponseData)
                {
                    // Do nothing
                };
            }
            
            WVJBHandler handler;
            if ([message objectForKey:@"handlerName"])
            {
                handler = [_messageHandlerDic objectForKey:[message objectForKey:@"handlerName"]];
                if (!handler)
                {
                    return responseCallback(@{});
                }
            }
            else
            {
                
            }
            
            @try
            {
                id data = [message objectForKey:@"data"];
                handler(data, responseCallback);
            }
            @catch (NSException *exception)
            {
                NSLog(@"BkJsBridge: WARNING: objc handler threw. %@ %@", message, exception);
            }
        }
    }
}

//序列化JSON
- (NSString *)serializeMessage:(id)message
{
    return [[NSString alloc] initWithData:[NSJSONSerialization dataWithJSONObject:message options:0 error:nil] encoding:NSUTF8StringEncoding];
}

//反序列化JSON
- (NSArray *)deserializeMessageJSON:(NSString *)messageJSON
{
    return [NSJSONSerialization JSONObjectWithData:[messageJSON dataUsingEncoding:NSUTF8StringEncoding] options:NSJSONReadingAllowFragments error:nil];
}

#pragma mark - UIWebViewDelegate
#pragma mark - UIWebViewDelegate
- (void)webViewDidFinishLoad:(UIWebView *)webView
{
    if (webView != _webView)
    {
        return;
    }
    
    _numRequestsLoading--;
    
    if (_numRequestsLoading == 0 && ![[webView stringByEvaluatingJavaScriptFromString:@"typeof BkJsBridge == 'object'"] isEqualToString:@"true"])
    {
        NSString *filePath = [[NSBundle mainBundle] pathForResource:@"BkJsBridge" ofType:@"js"];
        NSString *js = [NSString stringWithContentsOfFile:filePath encoding:NSUTF8StringEncoding error:nil];
        [webView stringByEvaluatingJavaScriptFromString:js];
    }
}

- (void)webView:(UIWebView *)webView didFailLoadWithError:(NSError *)error
{
    if (webView != _webView)
    {
        return;
    }
    
    _numRequestsLoading--;
}

- (BOOL)webView:(UIWebView *)webView shouldStartLoadWithRequest:(NSURLRequest *)request navigationType:(UIWebViewNavigationType)navigationType
{
    if (webView != _webView)
    {
        return YES;
    }
    
    NSURL *url = [request URL];
    if ([[url scheme] isEqualToString:kCustomProtocolScheme])
    {
        if ([[url host] isEqualToString:kQueueHasMessage])
        {
            [self refreshMessageQueue];
        }
        else
        {
            NSLog(@"BkJsBridge: WARNING: Received unknown BkJsBridge command %@://%@", kCustomProtocolScheme, [url path]);
        }
        
        return NO;
    }
    else
    {
        return YES;
    }
}

- (void)webViewDidStartLoad:(UIWebView *)webView
{
    if (webView != _webView)
    {
        return;
    }
    
    _numRequestsLoading++;
}

@end

File: JsBridgeDemo/WebViewJavascriptBridge/BkJsBridge.js
; (function () {
    if (window.BkJsBridge) { return }
    var messagingIframe
    var sendMessageQueue = []
    var receiveMessageQueue = []
    var messageHandlers = {}

    var CUSTOM_PROTOCOL_SCHEME = 'wvjbscheme'
    var QUEUE_HAS_MESSAGE = '__WVJB_QUEUE_MESSAGE__'

    var responseCallbacks = {}
    var uniqueId = 1

    //创建队列准备iframe
    function createQueueReadyIframe(doc) {
        messagingIframe = doc.createElement('iframe')
        messagingIframe.style.display = 'none'
        doc.documentElement.appendChild(messagingIframe)
    }

    function init(messageHandler) {
        if (BkJsBridge.messageHandler) {
            throw new Error('BkJsBridge.init called twice')
        }

        BkJsBridge.messageHandler = messageHandler
        var receivedMessages = receiveMessageQueue
        receiveMessageQueue = null

        for (var i = 0; i < receivedMessages.length; i++) {
            dispatchMessageFromObjC(receivedMessages[i])
        }
    }

    // js发送消息给oc(数据，回调方法)
    function send(data, responseCallback) {
        doSend({ data: data }, responseCallback)
    }

    // oc调js方法(方法名，回调方法)
    function registerHandler(handlerName, handler) {
    		if(handler.length == 0){
    				var innerHandler = function(data, responseCallback){
    					handler.call(this);
    					responseCallback();
    				};
    		}
    		else if(handler.length == 1){
    				innerHandler = function(data, responseCallback){
    					handler.call(this,data);
    					responseCallback();
    				};
    		}
    		else{
    				innerHandler = handler;
    		}
        messageHandlers[handlerName] = innerHandler;
    }

    //js调oc方法(方法名，数据，回调方法)
    function callHandler(handlerName, data, responseCallback) {
   alert("12345")
    		if(arguments.length == 1){
    				data = {};
    				responseCallback = function(result){
    				};
    		}
    		else if(arguments.length == 2){
    				responseCallback = function(result){
    				};
    		}
    		else{
    				
    		}
    		doSend({ handlerName: handlerName, data: data }, responseCallback)        
    }

    function doSend(message, responseCallback) {
        if (responseCallback) {
            var callbackId = 'cb_' + (uniqueId++) + '_' + new Date().getTime()
            responseCallbacks[callbackId] = responseCallback
            message['callbackId'] = callbackId
        }
        sendMessageQueue.push(message)
        messagingIframe.src = CUSTOM_PROTOCOL_SCHEME + '://' + QUEUE_HAS_MESSAGE
    }

    //获取队列
    function fetchQueue() {
        var messageQueueString = JSON.stringify(sendMessageQueue)
        sendMessageQueue = []
        return messageQueueString
    }

    //获取从oc发送来的消息
    function dispatchMessageFromObjC(messageJSON) {
        setTimeout(function timeoutDispatchMessageFromObjC() {
            var message = JSON.parse(messageJSON)
            var messageHandler

            if (message.responseId) {
                var responseCallback = responseCallbacks[message.responseId]
                if (!responseCallback) { return; }
                responseCallback(message.responseData)
                delete responseCallbacks[message.responseId]
            } else {
                var responseCallback
                if (message.callbackId) {
                    var callbackResponseId = message.callbackId
                    responseCallback = function (responseData) {
                        doSend({ responseId: callbackResponseId, responseData: responseData })
                    }
                }

                var handler = BkJsBridge.messageHandler
                if (message.handlerName) {
                    handler = messageHandlers[message.handlerName]
                }

                try {
                    handler(message.data, responseCallback)
                } catch (exception) {
                    if (typeof console != 'undefined') {
                        console.log("BkJsBridge: WARNING: javascript handler threw.", message, exception)
                    }
                }
            }
        })
    }

    //处理从oc发送来的消息
    function handleMessageFromObjC(messageJSON) {
        if (receiveMessageQueue) {
            receiveMessageQueue.push(messageJSON)
        } else {
            dispatchMessageFromObjC(messageJSON)
        }
    }

    window.BkJsBridge = {
        init: init,
        send: send,
        on: registerHandler,
        call: callHandler,
        fetchQueue: fetchQueue,
        handleMessageFromObjC: handleMessageFromObjC
    }

    var doc = document
    createQueueReadyIframe(doc)

    init(function (data, responseCallback) {
				responseCallback();
    });

    var readyEvent = doc.createEvent('Events')
    readyEvent.initEvent('BkJsBridgeReady')
    readyEvent.bridge = BkJsBridge
    doc.dispatchEvent(readyEvent)
})();
